walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,100 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  public static final Comparator<Event> SORT_BY_START = new Comparator<Event>() {
+    @Override
+    public int compare(Event a, Event b) {
+      return Long.compare(a.getWhen().start(), b.getWhen().start());
+    }
+  };
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) { 
+      Collection<TimeRange> expected = Arrays.asList();
+      return expected;
+    } else if (events.isEmpty() || request.getAttendees().isEmpty()) {
+      Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);
+      return expected;
+    } else {
+      // sort events by start time
+      List<Event> eventsList = new ArrayList<Event>();
+      for (Event event : events) {
+        if (request.getAttendees().containsAll(event.getAttendees())) {
+          eventsList.add(event);
+        }
+      }
+      Collections.sort(eventsList, SORT_BY_START);
+      
+      //ignore people not attending
+      if (eventsList.isEmpty()) {
+        Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);
+        return expected;
+      }
+
+      //merge event times
+      List<TimeRange> mergedEventsList = new ArrayList<TimeRange>();
+      if (eventsList.size() >= 2) {
+        for (int i = 0; i < eventsList.size()-1; i++) {
+          Event event1 = eventsList.get(i);
+          Event event2 = eventsList.get(i+1);
+          if (event1.getWhen().contains(event2.getWhen())) {
+            mergedEventsList.add(event1.getWhen());
+          } else if (event1.getWhen().overlaps(event2.getWhen())) {
+            mergedEventsList.add(TimeRange.fromStartEnd(event1.getWhen().start(), event2.getWhen().end(), false));
+          } else {
+            mergedEventsList.add(event1.getWhen());
+            mergedEventsList.add(event2.getWhen());
+          }
+        }
+      } else {
+        mergedEventsList.add(eventsList.get(0).getWhen());
```

Comment:
I'd either put this check before the complex logic, or add a comment reminding what it means if we're in this `else`. The first option is probably my preference since there's less risk of a comment getting out of date`

Reply:
sure

Reply:
https://github.com/gannerr/step/blob/b6c136f564e4ab96a9a02fefec07fd4101d635dc/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java#L63-L66